WalletWasabi/WabiSabi/Client/CoinJoinClient.cs

```diff
@@ -815,6 +813,28 @@ private void LogCoinJoinSummary(ImmutableArray<AliceClient> registeredAliceClien
 			}
 		}
 
+		double winnerAnonLoss = GetAnonLoss(winner);
+		while ((winner.Sum(x => x.Amount) > liquidityClue) && (winnerAnonLoss > MaxWeightedAnonLoss))
+		{
+			List<SmartCoin> bestReducedWinner = winner;
+			var bestAnonLoss = winnerAnonLoss;
+
+			foreach (SmartCoin coin in winner.Except(new[] { selectedNonPrivateCoin }))
+			{
+				var reducedWinner = winner.Except(new[] { coin });
+				var anonLoss = GetAnonLoss(reducedWinner);
+
+				if (anonLoss <= bestAnonLoss)
+				{
+					bestAnonLoss = anonLoss;
+					bestReducedWinner = reducedWinner.ToList();
+				}
```

Comment:
```suggestion
				if (anonLoss <= bestAnonLoss)
				{
					bestAnonLoss = anonLoss;
					bestReducedWinner = reducedWinner.ToList();
				}
				if (winner == bestReducedWinner)
			    {
				    break;
			    }
				
```
My thinking: Prevent the infinite loop, by checking if we couldn't reduce our `anonLoss` no further below our expected threshold. If so, use our remaining coins in the round.

Reply:
> as removing coins from the winner doesn't decrease the anonLoss below the threshold

I don't think this is the problem. You can always remove the coin with the highest anonymity, which decreases the anonymity loss.

The problem arises when `winner` contains exactly the same coins as `selectedNonPrivateCoin`, which means there is no coin that can be removed.

Anyway, the change you suggested does solve the problem.

Do you agree, @andrewkozlik?

Reply:
> Good changes, actually works better, but in cases when the MaxWeightedAnonLoss is low (as it's expected for some to be low) this can cause an infinite loop, as removing coins from the winner doesn't decrease the anonLoss below the threshold.

By the way, did you actually experienced the algorithm ends up in an endless loop? @Szpoti

Reply:
The reason why Ondřej is asking is because I think this should not enter an infinite loop.

Firstly `selectedNonPrivateCoin` is a single coin. There are two possible cases:
1. `winner` contains one coin. Then this coin is `selectedNonPrivateCoin` and `winnerAnonLoss` is 0, therefore the while-loop will end, because the second condition isn't satisfied (assuming that `MaxWeightedAnonLoss` is non-negative, which it certainly shouldn't be).
2. `winner` contains more than one coin. Then the foreach-loop will select one of the coins in `winner` which is not `selectedNonPrivateCoin` and it will remove it.

So either the while-cycle ends or a coin is removed. So the change proposed by @Szpoti is OK, but it seems to me that it's unreachable.

Reply:
Yes, we are running into an infinite loop in the case that all `GetAnonLoss(reducedWinner);` is higher than `winnerAnonLoss` I get that intuitively it should not happen, but it does.

During testing, here's a case where it happens: 
```
winner =
(Index, Amount, AnonymitySet)
(0, 0.04782969, 29),
(1, 0.11665525, 9),
(2, 0.04782969, 9)

winnerAnonLoss = 4.51

selectedNonPrivateCoin = coin with index 0

anonLoss without 1 = 10
anonLoss without 2 = 5.82
```
-> infinite loop

Reply:
Good job guys! We didn't realize that after the change in 346ed7cf77c743a13d3f6dd464a90666e1c9f01f the removal of coins is no longer monotonic in the anon loss...

Reply:
```
Failed to push to the remote repository. See the Output window for more details.
Pushing pr/8938
Pushing refs/heads/pr/8938
Remote: Permission to trezor/WalletWasabi.git denied to molnard.
Error encountered while pushing to the remote repository: Git failed with a fatal error.
Git failed with a fatal error.
unable to access 'https://github.com/trezor/WalletWasabi.git/': The requested URL returned error: 403
```

Created another PR https://github.com/zkSNACKs/WalletWasabi/pull/9555/files?diff=split&w=1